File: mRemoteV1/Connection/Connection.Protocol.PuttyBase.vb

Imports System.Threading
Imports mRemote3G.App
Imports mRemote3G.App.Info
Imports mRemote3G.Connection.PuttySession
Imports mRemote3G.Messages
Imports mRemote3G.My
Imports mRemote3G.Security
Imports mRemote3G.Tools

Namespace Connection

    Namespace Protocol
        Public Class PuttyBase
            Inherits Base

#Region "Constants"

            Private Const IDM_RECONF As Int32 = &H50 ' PuTTY Settings Menu ID

#End Region

#Region "Private Properties"

            Dim _isPuttyNg As Boolean

#End Region

#Region "Public Properties"

            Private _PuttyProtocol As Putty_Protocol

            Public Property PuttyProtocol As Putty_Protocol
                Get
                    Return Me._PuttyProtocol
                End Get
                Set
                    Me._PuttyProtocol = value
                End Set
            End Property

            Private _PuttySSHVersion As Putty_SSHVersion

            Public Property PuttySSHVersion As Putty_SSHVersion
                Get
                    Return Me._PuttySSHVersion
                End Get
                Set
                    Me._PuttySSHVersion = value
                End Set
            End Property

            Private _PuttyHandle As IntPtr

            Public Property PuttyHandle As IntPtr
                Get
                    Return Me._PuttyHandle
                End Get
                Set
                    Me._PuttyHandle = value
                End Set
            End Property

            Private _PuttyProcess As Process

            Public Property PuttyProcess As Process
                Get
                    Return Me._PuttyProcess
                End Get
                Set
                    Me._PuttyProcess = value
                End Set
            End Property

            Private Shared _PuttyPath As String

            Public Shared Property PuttyPath As String
                Get
                    Return _PuttyPath
                End Get
                Set
                    _PuttyPath = value
                End Set
            End Property

            Public ReadOnly Property Focused As Boolean
                Get
                    If NativeMethods.GetForegroundWindow() = PuttyHandle Then
                        Return True
                    Else
                        Return False
                    End If
                End Get
            End Property

#End Region

#Region "Private Events & Handlers"

            Private Sub ProcessExited(sender As Object, e As EventArgs)
                MyBase.Event_Closed(Me)
            End Sub

#End Region

#Region "Public Methods"

            Public Sub New()
            End Sub

            Public Overrides Function Connect() As Boolean
                Try
                    _isPuttyNg = (PuttyTypeDetector.GetPuttyType() = PuttyTypeDetector.PuttyType.PuttyNg)

                    PuttyProcess = New Process
                    With PuttyProcess.StartInfo
                        .UseShellExecute = False
                        .FileName = _PuttyPath

                        Dim arguments As New CommandLineArguments
                        arguments.EscapeForShell = False

                        arguments.Add("-load", InterfaceControl.Info.PuttySession)

                        If Not TypeOf InterfaceControl.Info Is PuttyInfo Then
                            arguments.Add("-" & _PuttyProtocol.ToString)

                            If _PuttyProtocol = Putty_Protocol.ssh Then
                                Dim username = ""
                                Dim password = ""

                                If Not String.IsNullOrEmpty(InterfaceControl.Info.Username) Then
                                    username = InterfaceControl.Info.Username
                                Else
                                    If MySettingsProperty.Settings.EmptyCredentials = "windows" Then
                                        username = Environment.UserName
                                    ElseIf MySettingsProperty.Settings.EmptyCredentials = "custom" Then
                                        username = MySettingsProperty.Settings.DefaultUsername
                                    End If
                                End If

                                If Not String.IsNullOrEmpty(InterfaceControl.Info.Password) Then
                                    password = InterfaceControl.Info.Password
                                Else
                                    If MySettingsProperty.Settings.EmptyCredentials = "custom" Then
                                        password = Crypt.Decrypt(MySettingsProperty.Settings.DefaultPassword,
                                                                 General.EncryptionKey)
                                    End If
                                End If

                                arguments.Add("-" & _PuttySSHVersion)

                                If Not ((Force And Info.Force.NoCredentials) = Info.Force.NoCredentials) Then
                                    If Not String.IsNullOrEmpty(username) Then
                                        arguments.Add("-l", username)
                                    End If
                                    If Not String.IsNullOrEmpty(password) Then
                                        arguments.Add("-pw", password)
                                    End If
                                End If
                            End If

                            arguments.Add("-P", InterfaceControl.Info.Port.ToString)
                            arguments.Add(InterfaceControl.Info.Hostname)
                        End If

                        If _isPuttyNg Then
                            arguments.Add("-hwndparent", InterfaceControl.Handle.ToString())
                        End If

                        .Arguments = arguments.ToString
                    End With

                    PuttyProcess.EnableRaisingEvents = True
                    AddHandler PuttyProcess.Exited, AddressOf ProcessExited

                    PuttyProcess.Start()
                    PuttyProcess.WaitForInputIdle(MySettingsProperty.Settings.MaxPuttyWaitTime*1000)

                    Dim startTicks As Integer = Environment.TickCount
                    While _
                        PuttyHandle.ToInt32 = 0 And
                        Environment.TickCount < startTicks + (MySettingsProperty.Settings.MaxPuttyWaitTime*1000)
                        If _isPuttyNg Then
                            PuttyHandle = NativeMethods.FindWindowEx(InterfaceControl.Handle, 0, vbNullString,
                                                                     vbNullString)
                        Else
                            PuttyProcess.Refresh()
                            PuttyHandle = PuttyProcess.MainWindowHandle
                        End If
                        If PuttyHandle.ToInt32 = 0 Then Thread.Sleep(0)
                    End While

                    If Not _isPuttyNg Then
                        NativeMethods.SetParent(PuttyHandle, InterfaceControl.Handle)
                    End If

                    Runtime.MessageCollector.AddMessage(MessageClass.InformationMsg, Language.Language.strPuttyStuff,
                                                        True)

                    Runtime.MessageCollector.AddMessage(MessageClass.InformationMsg,
                                                        String.Format(Language.Language.strPuttyHandle,
                                                                      PuttyHandle.ToString), True)
                    Runtime.MessageCollector.AddMessage(MessageClass.InformationMsg,
                                                        String.Format(Language.Language.strPuttyTitle,
                                                                      PuttyProcess.MainWindowTitle), True)
                    Runtime.MessageCollector.AddMessage(MessageClass.InformationMsg,
                                                        String.Format(Language.Language.strPuttyParentHandle,
                                                                      InterfaceControl.Parent.Handle.ToString), True)

                    Resize(Me, New EventArgs)

                    MyBase.Connect()
                    Return True
                Catch ex As Exception
                    Runtime.MessageCollector.AddMessage(MessageClass.ErrorMsg,
                                                        Language.Language.strPuttyConnectionFailed & vbNewLine &
                                                        ex.ToString())
                    Return False
                End Try
            End Function

            Public Overrides Sub Focus()
                Try
                    If ConnectionWindow.InTabDrag Then Return
                    NativeMethods.SetForegroundWindow(PuttyHandle)
                Catch ex As Exception
                    Runtime.MessageCollector.AddMessage(MessageClass.ErrorMsg,
                                                        Language.Language.strPuttyFocusFailed & vbNewLine &
                                                        ex.ToString(), True)
                End Try
            End Sub

            Public Overrides Sub Resize(sender As Object, e As EventArgs)
                Try
                    If InterfaceControl.Size = Size.Empty Then Return
                    NativeMethods.MoveWindow(PuttyHandle, -SystemInformation.FrameBorderSize.Width,
                                             -(SystemInformation.CaptionHeight + SystemInformation.FrameBorderSize.Height),
                                             InterfaceControl.Width + (SystemInformation.FrameBorderSize.Width * 2),
                                             InterfaceControl.Height + SystemInformation.CaptionHeight +
                                             (SystemInformation.FrameBorderSize.Height * 2), True)
                Catch ex As Exception
                    Runtime.MessageCollector.AddMessage(MessageClass.ErrorMsg,
                                                        Language.Language.strPuttyResizeFailed & vbNewLine &
                                                        ex.ToString(), True)
                End Try
            End Sub

            Public Overrides Sub Close()
                Try
                    If PuttyProcess.HasExited = False Then
                        PuttyProcess.Kill()
                    End If
                Catch ex As Exception
                    Runtime.MessageCollector.AddMessage(MessageClass.ErrorMsg,
                                                        Language.Language.strPuttyKillFailed & vbNewLine & ex.ToString(),
                                                        True)
                End Try

                Try
                    PuttyProcess.Dispose()
                Catch ex As Exception
                    Runtime.MessageCollector.AddMessage(MessageClass.ErrorMsg,
                                                        Language.Language.strPuttyDisposeFailed & vbNewLine &
                                                        ex.ToString(), True)
                End Try

                MyBase.Close()
            End Sub

            Public Sub ShowSettingsDialog()
                Try
                    NativeMethods.PostMessage(Me.PuttyHandle, NativeMethods.WM_SYSCOMMAND, IDM_RECONF, 0)
                    NativeMethods.SetForegroundWindow(Me.PuttyHandle)
                Catch ex As Exception
                    Runtime.MessageCollector.AddMessage(MessageClass.ErrorMsg,
                                                        Language.Language.strPuttyShowSettingsDialogFailed & vbNewLine &
                                                        ex.ToString(), True)
                End Try
            End Sub

#End Region

#Region "Enums"

            Public Enum Putty_Protocol
                ssh = 0
                telnet = 1
                rlogin = 2
                raw = 3
                serial = 4
            End Enum

            Public Enum Putty_SSHVersion
                None = 0
                ssh1 = 1
                ssh2 = 2
            End Enum

#End Region
        End Class
    End Namespace

End Namespace

File: mRemoteV1/CS/Connection/Connection.Protocol.PuttyBase.cs
// VBConversions Note: VB project level imports
using System.Collections.Generic;
using System;
using AxWFICALib;
using System.Drawing;
using System.Diagnostics;
using System.Data;
using AxMSTSCLib;
using Microsoft.VisualBasic;
using System.Collections;
using System.Windows.Forms;
// End of VB project level imports

using System.Runtime.InteropServices;
using System.ComponentModel;
using mRemoteNG.Messages;
//using mRemoteNG.App.Native;
using System.Threading;
using Microsoft.Win32;
//using mRemoteNG.App.Runtime;
using System.Text.RegularExpressions;
using mRemoteNG.Tools;


namespace mRemoteNG.Connection.Protocol
{
	public class PuttyBase : Base
	{
				
#region Constants
		private const int IDM_RECONF = 0x50; // PuTTY Settings Menu ID
#endregion
				
#region Private Properties
		bool _isPuttyNg;
#endregion
				
#region Public Properties
		private Putty_Protocol _PuttyProtocol;
public Putty_Protocol PuttyProtocol
		{
			get
			{
				return this._PuttyProtocol;
			}
			set
			{
				this._PuttyProtocol = value;
			}
		}
				
		private Putty_SSHVersion _PuttySSHVersion;
public Putty_SSHVersion PuttySSHVersion
		{
			get
			{
				return this._PuttySSHVersion;
			}
			set
			{
				this._PuttySSHVersion = value;
			}
		}
				
		private IntPtr _PuttyHandle;
public IntPtr PuttyHandle
		{
			get
			{
				return this._PuttyHandle;
			}
			set
			{
				this._PuttyHandle = value;
			}
		}
				
		private Process _PuttyProcess;
public Process PuttyProcess
		{
			get
			{
				return this._PuttyProcess;
			}
			set
			{
				this._PuttyProcess = value;
			}
		}
				
		private static string _PuttyPath;
public static string PuttyPath
		{
			get
			{
				return _PuttyPath;
			}
			set
			{
				_PuttyPath = value;
			}
		}
				
public bool Focused
		{
			get
			{
				if (GetForegroundWindow() == PuttyHandle)
				{
					return true;
				}
				else
				{
					return false;
				}
			}
		}
#endregion
				
#region Private Events & Handlers
		private void ProcessExited(object sender, System.EventArgs e)
		{
			base.EVENT_CLOSED(this);
		}
#endregion
				
#region Public Methods
		public PuttyBase()
		{
					
		}
				
		public override bool Connect()
		{
			try
			{
				_isPuttyNg = PuttyTypeDetector.GetPuttyType() == PuttyTypeDetector.PuttyType.PuttyNg;
						
				PuttyProcess = new Process();
				PuttyProcess.StartInfo.UseShellExecute = false;
				PuttyProcess.StartInfo.FileName = _PuttyPath;
						
				CommandLineArguments arguments = new CommandLineArguments();
				arguments.EscapeForShell = false;
						
				arguments.Add("-load", InterfaceControl.Info.PuttySession);
						
				if (!(InterfaceControl.Info is PuttySession.Info))
				{
					arguments.Add("-" + _PuttyProtocol.ToString());
							
					if (_PuttyProtocol == Putty_Protocol.ssh)
					{
						string username = "";
						string password = "";
								
						if (!string.IsNullOrEmpty(InterfaceControl.Info.Username))
						{
							username = InterfaceControl.Info.Username;
						}
						else
						{
							if (My.Settings.Default.EmptyCredentials == "windows")
							{
								username = Environment.UserName;
							}
							else if (My.Settings.Default.EmptyCredentials == "custom")
							{
								username = System.Convert.ToString(My.Settings.Default.DefaultUsername);
							}
						}
								
						if (!string.IsNullOrEmpty(InterfaceControl.Info.Password))
						{
							password = InterfaceControl.Info.Password;
						}
						else
						{
							if (My.Settings.Default.EmptyCredentials == "custom")
							{
								password = Security.Crypt.Decrypt(System.Convert.ToString(My.Settings.Default.DefaultPassword), App.Info.General.EncryptionKey);
							}
						}
								
						arguments.Add("-" + System.Convert.ToString(_PuttySSHVersion));
								
						if (!((Force & Info.Force.NoCredentials) == (int) Info.Force.NoCredentials))
						{
							if (!string.IsNullOrEmpty(username))
							{
								arguments.Add("-l", username);
							}
							if (!string.IsNullOrEmpty(password))
							{
								arguments.Add("-pw", password);
							}
						}
					}
							
					arguments.Add("-P", InterfaceControl.Info.Port.ToString());
					arguments.Add(InterfaceControl.Info.Hostname);
				}
						
				if (_isPuttyNg)
				{
					arguments.Add("-hwndparent", InterfaceControl.Handle.ToString());
				}
						
				PuttyProcess.StartInfo.Arguments = arguments.ToString();
						
				PuttyProcess.EnableRaisingEvents = true;
				PuttyProcess.Exited += ProcessExited;
						
				PuttyProcess.Start();
				PuttyProcess.WaitForInputIdle(System.Convert.ToInt32(My.Settings.Default.MaxPuttyWaitTime * 1000));
						
				int startTicks = Environment.TickCount;
				while (PuttyHandle.ToInt32() == 0 & Environment.TickCount < startTicks + (My.Settings.Default.MaxPuttyWaitTime * 1000))
				{
					if (_isPuttyNg)
					{
						PuttyHandle = FindWindowEx(InterfaceControl.Handle, 0, Constants.vbNullString, Constants.vbNullString);
					}
					else
					{
						PuttyProcess.Refresh();
						PuttyHandle = PuttyProcess.MainWindowHandle;
					}
					if (PuttyHandle.ToInt32() == 0)
					{
						Thread.Sleep(0);
					}
				}
						
				if (!_isPuttyNg)
				{
					SetParent(PuttyHandle, InterfaceControl.Handle);
				}
						
				MessageCollector.AddMessage(MessageClass.InformationMsg, My.Language.strPuttyStuff, true);
						
				MessageCollector.AddMessage(MessageClass.InformationMsg, string.Format(My.Language.strPuttyHandle, PuttyHandle.ToString()), true);
				MessageCollector.AddMessage(MessageClass.InformationMsg, string.Format(My.Language.strPuttyTitle, PuttyProcess.MainWindowTitle), true);
				MessageCollector.AddMessage(MessageClass.InformationMsg, string.Format(My.Language.strPuttyParentHandle, InterfaceControl.Parent.Handle.ToString()), true);
						
				Resize(this, new EventArgs());
						
				base.Connect();
				return true;
			}
			catch (Exception ex)
			{
				MessageCollector.AddMessage(MessageClass.ErrorMsg, My.Language.strPuttyConnectionFailed + Constants.vbNewLine + ex.Message);
				return false;
			}
		}
				
		public override void Focus()
		{
			try
			{
				if (ConnectionWindow.InTabDrag)
				{
					return ;
				}
				SetForegroundWindow(PuttyHandle);
			}
			catch (Exception ex)
			{
				MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg, My.Language.strPuttyFocusFailed + Constants.vbNewLine + ex.Message, true);
			}
		}
				
		public override void Resize(object sender, EventArgs e)
		{
			try
			{
				if (InterfaceControl.Size == Size.Empty)
				{
					return ;
				}
				MoveWindow(PuttyHandle, System.Convert.ToInt32(- SystemInformation.FrameBorderSize.Width), System.Convert.ToInt32(- (SystemInformation.CaptionHeight + SystemInformation.FrameBorderSize.Height)), InterfaceControl.Width + (SystemInformation.FrameBorderSize.Width * 2), InterfaceControl.Height + SystemInformation.CaptionHeight + (SystemInformation.FrameBorderSize.Height * 2), true);
			}
			catch (Exception ex)
			{
				MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg, My.Language.strPuttyResizeFailed + Constants.vbNewLine + ex.Message, true);
			}
		}
				
		public override void Close()
		{
			try
			{
				if (PuttyProcess.HasExited == false)
				{
					PuttyProcess.Kill();
				}
			}
			catch (Exception ex)
			{
				MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg, My.Language.strPuttyKillFailed + Constants.vbNewLine + ex.Message, true);
			}
					
			try
			{
				PuttyProcess.Dispose();
			}
			catch (Exception ex)
			{
				MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg, My.Language.strPuttyDisposeFailed + Constants.vbNewLine + ex.Message, true);
			}
					
			base.Close();
		}
				
		public void ShowSettingsDialog()
		{
			try
			{
				PostMessage(this.PuttyHandle, WM_SYSCOMMAND, IDM_RECONF, 0);
				SetForegroundWindow(this.PuttyHandle);
			}
			catch (Exception ex)
			{
				MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg, My.Language.strPuttyShowSettingsDialogFailed + Constants.vbNewLine + ex.Message, true);
			}
		}
#endregion
				
#region Enums
		public enum Putty_Protocol
		{
			ssh = 0,
			telnet = 1,
			rlogin = 2,
			raw = 3,
			serial = 4
		}
				
		public enum Putty_SSHVersion
		{
			ssh1 = 1,
			ssh2 = 2
		}
#endregion
	}
}
